### PR TITLE
[RFC] Autonomous schedules

### PR DESCRIPTION
**Pull request by attila-i-szabo**
We sometimes have hyperparameters we'd like to change during the training procedure, but they aren't learning rates (so they don't fit with full `optax` optimizers) nor is it convenient to store them in the object that uses them. The best example is `diag_shift` in QGT: literature suggests it's good to lower its value as SR goes on, but the QGT object lives for a single iteration, so it has no memory of such evolution. In #1060, we'd have double the problem, as the pseudo-Hessian object would have to tune both `diag_shift` and the learning rate that no longer fits in the `optax` paradigm.

This PR solves this problem by introducing a decorator that turns a schedule (i.e. a step count → parameter function) into an "autonomous schedule" (i.e. a function with internal state that can be called without arguments to return `schedule(0), schedule(1),...` on each call). E.g.:
```python
@nk.optimizer.autonomous
def schedule(x):
    return 3*x
schedule()
>>> 0
schedule()
>>> 3
schedule()
>>> 6
```

The QGT "smart constructors" are also modified to take such autonomous schedules for `diag_shift` in addition to numbers. If a function is passed to them, they call it without arguments to convert it into the desired value of `diag_shift` for that iteration. It would be nice to place this logic closer to the definition of `LinearOperator`, but I can't think of a way to do this with a `dataclass` (without breaking the contract with using `setattr`).

### CODE REVIEW

**Comment by PhilipVinc** [comment]
Nice and surely needed! 

However I'm not sure I like the approach and I think the design could be improved a bit, but should be discussed.

Motivation rises from the fact that you can't control how many times a QGT object will be constructed during an optimisation. Moreover, it might be possible that one day our whole `MCState` be differentiable (not only it's inner workings) and this would break it.

I'd really like to strive for having `QGT` objects be dump matrices, so that you should only feed them pure numbers.
The smart logic such as this should live in the preconditioner itself.
Maybe with a `preconditioner_state` much akin to how optax optimisers have a `optimiser state` object.
(Note: that would be breaking, because preconditions right now are 'pure' functions that are applied every step, while in this way we'd have to initialise them and then update the state at each step. 
With some smart we might be able to support both behaviours.
And I think I like the parallelism between `optimiser_state` and `preconditioner_state` )

**Comment by femtobit** [comment]
> (Note: that would be breaking, because preconditions right now are 'pure' functions that are applied every step, while in this way we'd have to initialise them and then update the state at each step.

I'd be very much in favor of keeping our preconditioners as pure functions. Isn't there a way to have them support schedules of the form `step -> parameter_value` directly, without adding hidden state?

**Comment by PhilipVinc** [comment]
On a second thought, I think they already have an hidden state because they are callable objects.

The issue is that the diag_shift is a property of the QGT(constructor) and not of SR, the preconditioned itself (or is it?) . 

It might make sense to make 'diag_shift' an internal field of QGT objects (you can always edit it manually with the addition operator) so that the preconditioner becomes responsible of the shift scheduling instead of the QGT itself.

**Comment by attila-i-szabo** [comment]
Yeah, `LinearPreconditioner` already has a lot of hidden state (e.g. the previous solution to be used as initial guess).

You're right that it would be a better place to keep track of the step count. What I'd like to avoid is drivers calculating the schedules of all sorts of hyperparameters, but rather them with the objects that use them (as optax optimisers do). 

On second thoughts, the "autonomous schedule" design is also not great because if you construct a QGT several times in one optimisation (as you would for RK time evolution?) then each of them will be different. 

Then, if we put it in the preconditioner (which does persist), it would be easier to store the `step count -> parameter` schedule and the step count explicitly and have a `forward()` function that increases the latter. We could even put that into `LinearPreconditioner` and subclasses can use it however they need to.

**Comment by femtobit** [comment]
> On second thoughts, the "autonomous schedule" design is also not great because if you construct a QGT several times in one optimisation (as you would for RK time evolution?) then each of them will be different.

I agree. Speaking of time evolution, it'd be great if the schedules would work there. Since the time evolution driver uses the current time as `step_value`, a mapping `step_value -> parameters` could be very useful. (`step_value` is equal to the step count in normal VMC, so it can be used in both cases.)

> Then, if we put the schedule in the preconditioner (which does persist), it would be easier to store the step count -> parameter schedule and the step count explicitly and have a forward() function that increases the latter. We could even put that into LinearPreconditioner and subclasses can use it however they need to.

That sounds good to me (using `step_value` instead of `step_count`).

**Comment by attila-i-szabo** [comment]
I suppose the natural way of supporting this then would be not having `step_value` as state but supplied to each call of the `LinearPreconditioner` and have drivers supply it?

**Comment by femtobit** [comment]
> I suppose the natural way of supporting this then would be not having step_value as state but supplied to each call of the `LinearPreconditioner` and have drivers supply it?

Yes, I think that would be ideal.

**Comment by PhilipVinc** [comment]
Can you mock the API that would result before implementing it?
What you are suggesting is that the preconditioner, instead of being a callable with signature
```
def preconditioner(vstate: VariationalState, gradient: PyTree) -> PyTree: 
```
becomes 
```
def preconditioner(vstate: VariationalState, gradient: PyTree, *, step_value:float) -> PyTree:
```
?

From a point of view, it's nice.
From another, it seems arbitrary that we now might pass `step_value`. What about other stuff? `energy: Stats` for example?

--

Another issue: I guess we'd now be adding a new field to `SR(diag_shift: Union[Schedule, Number]=default_value)`. What about the existing way to define the `diag_shift` through the QGT object?

--

@femtobit anyhow, this machinery is not used by `TDVP` in any way. We only supply the linear solver and the qgt...

**Comment by femtobit** [comment]
> @femtobit anyhow, this machinery is not used by TDVP in any way. We only supply the linear solver and the qgt...

That is true right now, but could be changed if it is worth it.

> `def preconditioner(vstate: VariationalState, gradient: PyTree, *, step_value:float) -> PyTree:`

You mean `LinearPreconditioner.__call__` here? Still, this requires some more consideration since the precoditoner just uses what you called `lhs_constructor`.

This has the type
```python
LHSConstructorT = Callable[[VariationalState], LinearOperator]
"""Signature for the constructor of a LinerOperator"""
```
Should the additional parameters go there? If `LHSConstructorT` accepted additional arguments, one could pass something like
```python
lhs_cons = lambda vstate, step_value: return QGTJacobianPytree(vstate, diag_shift=schedule(step_value)
```
Its not clear to me yet, how to repackage this so users can provide the schedule conveniently. (`QGTJacobianDense(..., diag_shift=schedule)` seems problematic unless we want to pass the `step_value` through to both `lhs_constructor` and `solver`.

**Comment by PhilipVinc** [comment]
>You mean LinearPreconditioner.__call__ here? Still, this requires some more consideration since the precoditoner just uses what you called lhs_constructor.

Yes. Because a preconditioner is documented [here](https://www.netket.org/docs/custom_preconditioners.html) to be a callable with that signature.

>Its not clear to me yet, how to repackage this so users can provide the schedule conveniently. (QGTJacobianDense(..., diag_shift=schedule) seems problematic unless we want to pass the step_value through to both lhs_constructor and solver.

To me a good API would be

```python
preconditioner=SR(qgt=nk.optimizer.qgt.thisType, solver=..., diag_shift=schedule)
```

**Comment by attila-i-szabo** [comment]
> becomes
> 
> ```
> def preconditioner(vstate: VariationalState, gradient: PyTree, *, step_value:float) -> PyTree:
> ```
> 
> ?

Yes that's what I had in mind

> From a point of view, it's nice.
> From another, it seems arbitrary that we now might pass `step_value`. What about other stuff? `energy: Stats` for example?

I'm not against that either if there's a use case for it, but there's none yet. (#1060 uses more elaborate things, so passing stuff like this wouldn't help... we probably can't spare a new driver class there)

> Another issue: I guess we'd now be adding a new field to `SR(diag_shift: Union[Schedule, Number]=default_value)`. What about the existing way to define the `diag_shift` through the QGT object?

We can make this `diag_shift` optional and None by default so if you pass it through the QGT it wouldn't affect that.

> Should the additional parameters go there? If `LHSConstructorT` accepted additional arguments, one could pass something like

I think the problem is more how the driver knows what to pass...

**Comment by PhilipVinc** [comment]
>>Should the additional parameters go there? If LHSConstructorT accepted additional arguments, one could pass something like
>I think the problem is more how the driver knows what to pass...

I don't see the problem? If we change the signature then we just change `LHSConstructorT` as well?

For backward compatibility we can do a 'best-effort' attempt using inspect and checking that if the preconditioner passed has 2 arguments, we drop the new one + issue a deprecation warning..?

**Comment by attila-i-szabo** [comment]
> I don't see the problem? If we change the signature then we just change `LHSConstructorT` as well?

So would the idea be that we change to
```python3
LHSConstructorT = Callable[[VariationalState, Number], LinearOperator]
```
where the `Number` is for `step_value` and change https://github.com/netket/netket/blob/bdf7071ff4000a31adce360a86f132708fc6db82/netket/optimizer/sr/api.py#L178-L182 to 
```python
    return SR(
        lambda vs, step_value: qgt(vs, diag_shift=diag_shift(step_value), **kwargs),
        solver=solver,
        solver_restart=solver_restart,
    )
```
(or rather some conditional depending on whether `diag_shift` is a number or a schedule)? Maybe it would be more elegant to move this implementation in the `SR` class but I guess this works fine.

> For backward compatibility we can do a 'best-effort' attempt using inspect and checking that if the preconditioner passed has 2 arguments, we drop the new one + issue a deprecation warning..?

What do you mean the preconditioner has 2 args? 
`LinearPreconditioner.__call__` is called from the driver only, so I wouldn't worry about that (I guess we can add a default value of `step_value` to make sure stuff doesn't break in users' code).

If you mean supplying `diag_shift` through the QGT and making your `LinearPreconditioner` by hand, it would also be fine to have a code path that keeps it working, though I'm not sure how we'd check how many arguments a function is happy to take...

**Comment by attila-i-szabo** [comment]
> > From a point of view, it's nice.
> > From another, it seems arbitrary that we now might pass `step_value`. What about other stuff? `energy: Stats` for example?
> 
> I'm not against that either if there's a use case for it, but there's none yet. (#1060 uses more elaborate things, so passing stuff like this wouldn't help... we probably can't spare a new driver class there)

I thought I'd write a more principled take on this.

You're right that preconditioners could ask for all sorts of things that would be convenient to supply this. However, most of these (like `energy: Stats`) would be functions of `vstate`, so if it speeds things up to supply them, they should be added to `vstate` in some way. 

On the other hand, `vstate` is a snapshot of the ansatz at some point in the training, so it wouldn't make much sense to keep track of `step_value` in it, that belongs more to the driver, so it needs to be supplied separately from `vstate`. Also, it's generic, because all drivers I can think of would have some notion of epochs, but many of them don't need `energy` for example.

**Comment by attila-i-szabo** [comment]
BTW, could we call `step_value` `epoch`? I think it's just as generic, it sounds better, and `step_value` could also mean step size rather than where you are in the training...

**Comment by attila-i-szabo** [comment]
> Yes. I'm unconvinced if this `Number` should be a mandatory positional argument or an optional kw argument.
> 
> Same with `PreconditionerT = Callable[[VariationalState, PyTree, Number], PyTree]` should it be mandatory or optional?

Good question. I'd make it a kwarg at least for `PreconditionerT`, see below why.

> > What do you mean the preconditioner has 2 args?
> > LinearPreconditioner.**call** is called from the driver only, so I wouldn't worry about that (I guess we can add a default value of step_value to make sure stuff doesn't break in users' code).
> 
> Users can define their own preconditioners, and if you now use the 3-argument interface those will not work anymore.

I see. 
I thought about the opposite problem, our preconditioners (with a 3-argument interface) not working with users' drivers that only supply 2 arguments. That is easy to fix if you make `step_value` a kwarg with a default like `step_value=0`.
This problem would probably be breaking, if the preconditioner doesn't expect a third argument, it will crash, whether it's positional or kw.

> > BTW, could we call step_value epoch? I think it's just as generic, it sounds better, and step_value could also mean step size rather than where you are in the training...
> 
> Can you open another issue about this? I'm not against but to avoid changing the api all the time we can defer this to a breaking release later on..

We don't use either of these names anywhere yet, do we?

**Comment by femtobit** [comment]
> We don't use either of these names anywhere yet, do we?

They are public properties of the driver, so renaming them is breaking to user code. Also, in most cases (VMC, QSR, etc.) `step_value` would be the same as `step_count`. Only t-VMC is different and there machine learning terms such as `epoch` also seem out-of-place to me.

**Comment by attila-i-szabo** [comment]
> > We don't use either of these names anywhere yet, do we?
> 
> They are public properties of the driver, so renaming them is breaking to user code. Also, in most cases (VMC, QSR, etc.) `step_value` would be the same as `step_count`. Only t-VMC is different and there machine learning terms such as `epoch` also seem out-of-place to me.

I see, I haven't realised this was a property of the driver. Then sure, I wouldn't change its name. OTOH, what we call this in the preconditioner doesn't have to match the driver, so `step_value` could be used without breaking anything

**Comment by gcarleo** [comment]
I guess we could use this logic to allow also the number of samples to change with a schedule?

**Comment by PhilipVinc** [comment]
> even though avoiding recompiling at each change might be slightly tricky]

impossible for now

**Comment by PhilipVinc** [comment]
@attila-i-szabo I think we agree on the design. Do you have time to work on it?

**Comment by attila-i-szabo** [comment]
> > even though avoiding recompiling at each change might be slightly tricky]
> 
> impossible for now

Yeah, that would require jitting to work across different sized arrays, which is not the case atm. OTOH, we can try to cover that too, with a caveat to the user that they shouldn't change it often

> @attila-i-szabo I think we agree on the design. Do you have time to work on it?

Not a lot atm, but I can try to find some. Just so I'm clear about the design: 
* `PreconditionerT = Callable[[VariationalState, PyTree, Number], PyTree]`, where the `Number` is a kwarg for `step_value` (BTW, how do you express that in the type declaration?)
* `LHSConstructorT = Callable[[VariationalState, Number], LinearOperator]` ditto
* drivers pass `step_value` into the preconditioner, which is a breaking change for user-defined preconditioners, albeit one that is trivial to fix by adding a `step_value` kwarg doing nothing
* `SR` is changed to handle `diag_shift` schedules (and internally convert a constant `diag_shift` into a schedule as well)

**Comment by attila-i-szabo** [comment]
Closed in favour of #1142
